### PR TITLE
Cs author profile

### LINKED ISSUE #151: cross server get author crashes when there is a friend that is not from our server
this might imply every endpoint that potentially needs to get author profile details to be fixed to handle when the author is not from our server
ie: 
GET `/api/author/<authorid>`
GET `/api/followers/<authorid>`
and potentially more

### PR DESCRIPTION
**Pull request by hiufungk**
resolves #151 
image
 can view author profile from foreign server

### CODE REVIEW

File: backend/api/view/AuthorProfileView.py [hunk 1]
```diff
@@ -13,28 +19,29 @@ class AuthorProfileView(generics.GenericAPIView):
 
     def is_mutable(self, key):
         for ele in self.mutable_keys:
-            if(key == ele):
+            if (key == ele):
```
Comment (htruong4) [inline backend/api/view/AuthorProfileView.py]: You can probably just check if the key is in self.mutable_keys instead. Same thing as you're doing here but a nice one liner

File: backend/api/view/AuthorProfileView.py [hunk 2]
```diff
@@ -50,29 +57,90 @@ def post(self, request, uid):
...
+                uuid.UUID(uid)
+            except ValueError:
+                is_local_uuid = False
+            if is_local_uuid:
+                try:
+                    author_profile = AuthorProfile.objects.get(id=authorId)
+                    response_data = AuthorProfileSerializer(author_profile).data
+                    friends = Follow.objects.filter(authorA=response_data["id"], status="FRIENDS")
+                    friends_list_data = []
+                    for ele in friends:
```
Comment (htruong4) [inline backend/api/view/AuthorProfileView.py]: not too picky about this but i like for friend in friends better becasue it is more meaningful

File: frontend/src/pages/Author.jsx [hunk 3]
```diff
@@ -41,11 +41,52 @@ class Author extends Component {
         this.getFollowButton = this.getFollowButton.bind(this);
         this.sendUnfollowRequest = this.sendUnfollowRequest.bind(this);
         this.getFollowStatus = this.getFollowStatus.bind(this);
+        // this.getTabPanes = this.getTabPanes.bind(this);
```
Comment (htruong4) [inline frontend/src/pages/Author.jsx]: is this going to be used in the future? The commented out line that is

File: backend/api/view/AuthorProfileView.py (hunk 1 above)
Comment (hiufungk) [reply backend/api/view/AuthorProfileView.py]: you wrote this... i didnt change anything.......

File: backend/api/view/AuthorProfileView.py [hunk 4]
```diff
@@ -50,29 +54,91 @@ def post(self, request, uid):
...
-            friends_list_data = []
-            for ele in friends:
-                friend_fulll_id = ele.authorB
-                tmp = friend_fulll_id.split("author/")
-                host = tmp[0]
-                short_id = tmp[1]
-                # todo: check if host belongs to our server, call cross server endpoint if doesnt
-                friend_profile = AuthorProfile.objects.get(id=short_id)
-                serialized_author_profile = AuthorProfileSerializer(friend_profile)
+                                print(server_user)
```
Comment (htruong4) [inline backend/api/view/AuthorProfileView.py]: remove this printy boi or @shchau  will haunt you lmao

File: backend/api/view/StreamPostsView.py [hunk 5]
```diff
@@ -76,6 +77,7 @@ def get(self, request):
                     return Response("Get request fail, bad foreign host",
                                     status.HTTP_400_BAD_REQUEST)
                 except Exception as e:
+                    print("inside big exception")
```
Comment (htruong4) [inline backend/api/view/StreamPostsView.py]: remove this printy boi

**Review by htruong4 — CHANGES_REQUESTED** [review]

**Review by htruong4 — APPROVED** [review]

File: backend/api/view/AuthorProfileView.py [hunk 6]
```diff
@@ -50,29 +54,86 @@ def post(self, request, uid):
...
+            except ValueError:
+                is_local_uuid = False
+            if is_local_uuid:
+                try:
+                    author_profile = AuthorProfile.objects.get(id=authorId)
+                    response_data = AuthorProfileSerializer(author_profile).data
+                    friends = Follow.objects.filter(authorA=response_data["id"], status="FRIENDS")
+                    friends_list_data = []
+                    for friend in friends:
+                        friend_fulll_id = friend.authorB
```
Comment (derrick-w) [inline backend/api/view/AuthorProfileView.py]:
```suggestion
                        friend_full_id = friend.authorB
```

File: backend/api/endpoints.py [hunk 7]
```diff
@@ -9,14 +9,16 @@
...
+from django.urls import path
 
 urlpatterns = [
     url("^auth/register/$", RegistrationView.as_view()),
     url("^auth/login/$", LoginView.as_view()),
     url(r"^author/posts/?$", StreamPostsView.as_view()),
     url(r'^author/(?P<authorid>.*)/posts/?$', GetPostsView.as_view()),
     url(r'^author/(?P<authorid>.*)/friends/?$', CheckFriendsView.as_view()),
-    url(r"^author/(?P<uid>.*)$", AuthorProfileView.as_view()),
+    # url(r"^author/(?P<uid>.*)$", AuthorProfileView.as_view()),
```
Comment (derrick-w) [inline backend/api/endpoints.py:20]: remove comment?

File: backend/api/view/AuthorProfileView.py [hunk 8]
```diff
@@ -50,29 +54,86 @@ def post(self, request, uid):
...
+        elif server_user_exists:
+            try:
+                author_profile = AuthorProfile.objects.get(id=authorId)
+                response_data = AuthorProfileSerializer(author_profile).data
+                return Response(response_data, status.HTTP_200_OK)
+            except:
+                return Response("Author does not exist", status.HTTP_400_BAD_REQUEST)
         else:
-            return Response("Author does not exist", status.HTTP_400_BAD_REQUEST)
+            return Response("Request not from invalid place", status.HTTP_400_BAD_REQUEST)
```
Comment (derrick-w) [inline backend/api/view/AuthorProfileView.py]:
```suggestion
            return Response("Request not from valid server", status.HTTP_400_BAD_REQUEST)
```

File: frontend/src/pages/Author.jsx [hunk 9]
```diff
@@ -41,11 +41,51 @@ class Author extends Component {
...
+            currentPageAuthorId = decodeURIComponent(this.props.match.params.authorId);
+
+        HTTPFetchUtil.getRequest(hostUrl, true)
+            .then((httpResponse) => {
+                if (httpResponse.status === 200) {
+                    httpResponse.json().then((results) => {
+                        if (results.friends) {
+                            for (let i = 0; i < results.friends.length; i++) {
+                                if (results.friends[i].id === currentPageAuthorId){
+                                    this.setState({canFollow: false});
```
Comment (derrick-w) [inline frontend/src/pages/Author.jsx:67]: remember to remove all references to isFollowing and isSelf if deprecated

File: frontend/src/pages/Author.jsx [hunk 10]
```diff
@@ -54,14 +94,20 @@ class Author extends Component {
...
+                        if (results.friends) {
+                            for (let i = 0; i < results.friends.length; i++) {
+                                if (results.friends[i].id === authenticatedAuthorId){
+                                    canFollow = false;
+                                    break;
+                                }
                             }
+                        } else {
+                            // results.friends not available means it is a foreign friend
+                            // then we need to fetch our own profile to check if i am friend/following him already
```
Comment (derrick-w) [inline frontend/src/pages/Author.jsx]:
```suggestion
                            // then we need to fetch our own profile to check if I am friend/following him already
```
capitalization heeheexd

**Review by derrick-w — CHANGES_REQUESTED** [review]

**Review by derrick-w — APPROVED** [review]

**Pull request merged**